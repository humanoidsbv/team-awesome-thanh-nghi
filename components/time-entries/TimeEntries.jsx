import React from 'react';
import PropTypes from 'prop-types';

import './time-entries.scss';
import TimeEntry from '../time-entry';

// TODO: put this logic in a service -> generic and reusable
function dateToLocaleString(date) {
  const newDate = new Date(date);
  const options = { day: '2-digit', month: '2-digit', year: 'numeric' };
  const localeDate = newDate.toLocaleString('nl-NL', options);
  return localeDate;
}

const TimeEntries = ({ timeEntries }) => (
  <React.Fragment>
    {timeEntries.map(({
      client, date, endTime, id, startTime
    }, i) => {
      const localizeDate = dateToLocaleString(date);
      return (
        <React.Fragment>
          {(i === 0 || date !== timeEntries[i - 1].date)
            && <h2 className="entry-date">{localizeDate}</h2>
          }
          <TimeEntry
            client={client}
            endTime={endTime}
            key={id}
            startTime={startTime}
          />
        </React.Fragment>
      );
    })}
  </React.Fragment>
);

TimeEntries.propTypes = {
  timeEntries: PropTypes.arrayOf(PropTypes.shape({})).isRequired
};

export default TimeEntries;
