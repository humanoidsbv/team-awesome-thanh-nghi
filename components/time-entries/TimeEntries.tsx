import React from 'react';

import './time-entries.scss';

import Container from '../../shared/components/Container';
import NewTimeEntry from '../new-time-entry';
import PageHeader from '../../shared/components/PageHeader';
import Select from '../../shared/components/Select';
import TimeEntry from '../time-entry';
import { TimeEntryModel, TimeEntriesState } from '../../ducks/time-entries';
import { dateToLocaleString, timeToLocaleString } from '../../shared/services/converter-time/converter-time';

interface TimeEntriesProps {
  clients: string[];
  filterTimeEntries: Function;
  onDelete: Function;
  timeEntries: TimeEntryModel[];
}

class TimeEntries extends React.Component<TimeEntriesProps, TimeEntriesState> {
  handleChange = ({ target }) => {
    const { filterTimeEntries } = this.props;
    filterTimeEntries(target.value);
  }

  render() {
    const { clients, onDelete, timeEntries } = this.props;
    const count = timeEntries.length;

    return (
      <React.Fragment>
        <PageHeader
          title="Timesheets"
          subtitle={`${count} ${count === 1 ? 'Entry' : 'Entries'}`}
        />
        <Container>
          <NewTimeEntry />
          <div className="time-entries">
            <h2 className="time-entries__title">
              Time Entries
            </h2>
            <Select
              className="select-element__filter"
              defaultValue={{ label: 'All clients', value: '' }}
              onChange={this.handleChange}
              options={clients}
            />
          </div>
          {timeEntries.map(({
            clientName, endTime, id, startTime
          }, i) => {
            const localizeDate = dateToLocaleString(startTime);
            const localizeEndTime = timeToLocaleString(endTime);
            const localizeStartTime = timeToLocaleString(startTime);

            return (
              <React.Fragment key={id}>
                {(i === 0 || localizeDate !== dateToLocaleString(timeEntries[i - 1].startTime))
                  && <h2 className="entry-date">{localizeDate}</h2>
                }
                <TimeEntry
                  client={clientName}
                  endTime={localizeEndTime}
                  id={id}
                  onDelete={onDelete}
                  startTime={localizeStartTime}
                />
              </React.Fragment>
            );
          })}
        </Container>
      </React.Fragment>
    );
  }
}

export default TimeEntries;
