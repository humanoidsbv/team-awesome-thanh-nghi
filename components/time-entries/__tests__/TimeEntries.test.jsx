import React from 'react';
import toJson from 'enzyme-to-json';
import { shallow } from 'enzyme';
import TimeEntries from '../TimeEntries.tsx';

const deleteTimeEntry = () => {};
const filterTimeEntries = () => {};
const retrieveTimeEntries = () => {};

const timeEntriesSelector = [
  {
    activity: 'front-end',
    client: '3',
    clientName: 'Humanoids',
    endTime: '2019-03-20T16:00:00.000Z',
    id: '123',
    startTime: '2019-03-20T08:00:00.000Z'
  },
  {
    activity: 'design',
    client: '1',
    clientName: 'Port of Rotterdam',
    endTime: '2019-03-22T16:00:00.000Z',
    id: '124',
    startTime: '2019-03-20T08:00:00.000Z'
  },
  {
    activity: 'lol',
    client: '2',
    clientName: 'Bank of Amsterdam',
    endTime: '2019-03-25T16:00:00.000Z',
    id: '125',
    startTime: '2019-03-20T08:00:00.000Z'
  }
];

const clientNameIdSelector = [
  {
    label: 'Port of Rotterdam',
    value: '1'
  },
  {
    label: 'Port of Rotterdam',
    value: '2'
  },
  {
    label: 'Humanoids',
    value: '3'
  }
];

it('renders a snapshot', () => {
  const wrapper = shallow(
    <TimeEntries
      clients={clientNameIdSelector}
      filterTimeEntries={filterTimeEntries}
      onDelete={deleteTimeEntry}
      onRetrieve={retrieveTimeEntries}
      timeEntries={timeEntriesSelector}
    />
  );
  expect(toJson(wrapper)).toMatchSnapshot();
});
