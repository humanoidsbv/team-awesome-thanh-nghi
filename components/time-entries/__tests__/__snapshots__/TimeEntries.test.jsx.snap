// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a snapshot 1`] = `
<Fragment>
  <PageHeader
    subtitle="3 Entries"
    title="Timesheets"
  />
  <Container>
    <Connect(NewTimeEntryContainer) />
    <div
      className="time-entries"
    >
      <h2
        className="time-entries__title"
      >
        Time Entries
      </h2>
      <Select
        className="select-element__filter"
        defaultValue={
          Object {
            "label": "All clients",
            "value": "",
          }
        }
        id=""
        isInvalid={false}
        name=""
        onBlur={[Function]}
        onChange={[Function]}
        options={
          Array [
            Object {
              "label": "Port of Rotterdam",
              "value": "1",
            },
            Object {
              "label": "Port of Rotterdam",
              "value": "2",
            },
            Object {
              "label": "Humanoids",
              "value": "3",
            },
          ]
        }
        required={false}
      />
    </div>
    <h2
      className="entry-date"
    >
      20-03-2019
    </h2>
    <TimeEntry
      client="Humanoids"
      endTime="17:00"
      id="123"
      onDelete={[Function]}
      startTime="09:00"
    />
    <TimeEntry
      client="Port of Rotterdam"
      endTime="17:00"
      id="124"
      onDelete={[Function]}
      startTime="09:00"
    />
    <TimeEntry
      client="Bank of Amsterdam"
      endTime="17:00"
      id="125"
      onDelete={[Function]}
      startTime="09:00"
    />
  </Container>
</Fragment>
`;
